### ISSUE TITLE
Unable to login using the imported MD5 identity.

### ISSUE THREAD

**Issue by graysontvbs** [ory/kratos#3310]
### Preflight checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/kratos/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/kratos/blob/master/CONTRIBUTING.md).
- [x] This issue affects my [Ory Network](https://www.ory.sh/) project.
- [x] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [ ] I am signed up to the [Ory Security Patch Newsletter](https://ory.us10.list-manage.com/subscribe?u=ffb1a878e4ec6c0ed312a3480&id=f605a41b53).

### Describe the bug

I learned from the official source that it is possible to import identities using the MD5 encryption method, and Kratos performs re-hashing upon login. However, when I tested it, I couldn't log in using the imported accounts. After examining the source code, I discovered that this is due to a difference in the way the base64 parameter is passed.

Also the unit test cases for this [PR](https://github.com/ory/kratos/pull/2725) is incorrect. Taking the first test case as an example, after encrypting with md5 and base64, the expected result should be `MDk4ZjZiY2Q0NjIxZDM3M2NhZGU0ZTgzMjYyN2I0ZjY=`, but the test case uses `CY9rzUYh03PK3k6DJie09g==`.

The reason for this is that the encryption process involves passing the `byte[:]` result of md5 into the base64 encode function, rather than passing the string result of md5. If you use the following sample code, you can clearly see that the base64 output will be different depending on the input data:

```go
md5Hash := md5.Sum([]byte("test")) // 098f6bcd4621d373cade4e832627b4f6

//use byte[:] to base64 encode (X)
base64Encoded := base64.StdEncoding.EncodeToString(md5Hash[:])
fmt.Println(base64Encoded) //CY9rzUYh03PK3k6DJie09g==

//use string to base64 encode (O)
base64Encoded1 := base64.StdEncoding.EncodeToString([]byte(hex.EncodeToString(md5Hash[:])))
fmt.Println(base64Encoded1) //MDk4ZjZiY2Q0NjIxZDM3M2NhZGU0ZTgzMjYyN2I0ZjY=
```

ps. also I noted at the origin [pull request](https://github.com/ory/kratos/pull/2725#issuecomment-1584187695).

### Reproducing the bug

I run in my docker，here is the version what I use.

* docker image: kratos:v0.13.0

* identity json file 
```
{
    "schema_id": "default",
    "traits": {
        "email": "example@test.com.tw",
        "name": {"last_name": "user1", "first_name": "user1"}
    },
    "credentials": {
        "password": {
          "config": {
            "hashed_password": "$md5$MDk4ZjZiY2Q0NjIxZDM3M2NhZGU0ZTgzMjYyN2I0ZjY="
          }
        }
      }
}
```

* command 
```
kratos import identities  -e http://127.0.0.1:4433  user.yaml
```

### Version

v0.13.0

### On which operating system are you observing this issue?

macOS

### In which environment are you deploying?

Docker

**Comment by kmherrmann** [comment]
Did you manage to run the import successfully in the mean time, or are there still issues?

**Comment by graysontvbs** [comment]
Yes, I successfully imported the identities, but I still cannot log in. That's why I delved deeper into investigating this issue.

**Comment by Renkas** [comment]
I have same issue. Can't go forward with the migration to Kratos like this.

Using v1.3.1

Is there a workaround?

**Comment by graysontvbs** [comment]
I think they've already fixed it in this [PR](https://github.com/ory/kratos/pull/2725).

**Comment by Renkas** [comment]
I solved this by using migrate_hook.

But this bug cost us a lot of dev time ...

**Comment by graysontvbs** [comment]
Me too ...

**Comment by jonas-jonas** [comment]
I looked into this briefly. The issue is that we're passing the byte value encoded as base64 directly, e.g. `base64(md5("password"))`, to the comparison function. The expectation here was that we would use the hex encoded, encoded as base64, e.g. `base64(hex(md5("password")`.

AFAICT, [PHP's `md5` function implements the hex based approach](https://www.php.net/manual/en/function.md5.php), which is where the confusion is coming from, I assume. Other implementations seem to be generating the hash without first encoding the md5 digest as hex.

I believe we can try to decode the hex encoded hash during the comparison and silently fall back to the current behavior if that fails. This way we accommodate both approaches.

**Comment by hperl** [comment]
Thanks for the report. There are many ways to hash passwords, and this issue comes down to how you want to write the password hash.

In the docs about [importing passwords](https://www.ory.sh/docs/kratos/manage-identities/import-user-accounts-identities#importing-credentials) for MD5 we write

> Parameters
hash(string): The computed hash by the MD5 algorithm encoded to Base64.

So the correct way to compute the hash is indeed `base64(md5(password))`, not `base64(hex(md5(password)))`. This also makes sense, because both `base64` and `hex` are functions that transform bytes into ASCII, so chaining them is not really necessary.

So, what to do if you have your passwords as `hash = base64(hex(md5(password)))`? Simple, just run `base64_encode(hex_decode(base64_decode(hash)))` for each hash before you import, essentially getting rid of the unnecessary `hex` encoding.
